vault/raft.go

```diff
@@ -947,39 +947,28 @@ func (c *Core) JoinRaftCluster(ctx context.Context, leaderInfos []*raft.LeaderJo
 				}
 
 			case leaderInfo.AutoJoin != "":
+				// go-discover returns IPs
 				addrs, err := disco.Addrs(leaderInfo.AutoJoin, c.logger.StandardLogger(nil))
 				if err != nil {
 					c.logger.Error("failed to parse addresses from auto-join metadata", "error", err)
 				}
-
+				scheme := leaderInfo.AutoJoinScheme
+				if scheme == "" {
+					// default to HTTPS when no scheme is provided
+					scheme = "https"
+				}
+				port := leaderInfo.AutoJoinPort
+				if port == 0 {
+					// default to 8200 when no port is provided
+					port = 8200
+				}
 				for _, addr := range addrs {
-					u, err := url.Parse(addr)
-					if err != nil {
-						c.logger.Error("failed to parse discovered address", "error", err)
-						continue
-					}
-
-					if u.Scheme == "" {
-						scheme := leaderInfo.AutoJoinScheme
-						if scheme == "" {
-							// default to HTTPS when no scheme is provided
-							scheme = "https"
-						}
-
-						addr = fmt.Sprintf("%s://%s", scheme, addr)
+					if strings.Count(addr, ":") != 0 {
```

Comment:
There's an [SO post](https://stackoverflow.com/questions/22751035/golang-distinguish-ipv4-ipv6) that seems to indicate to correctly validate IPv6 vs IPv4 addresses we should test for `>=2` as an IPv4 address could contain a port. I'm not sure if it's worth being pedantic about the test here since discovery's API states it returns IPs. But since it's name is `Addrs` ... I can see a future change in the library breaking this in one direction or another.

I don't know how defensive you'd like to write these bits.

Reply:
Hmm, I think I'd honestly be fine with either, cause if go-discover starts returning ips with ports, then `u := fmt.Sprintf("%s://%s:%d", scheme, addr, port)` would also break. That being said, I'd still vote for `>=2` since that seems more correct.

Reply:
Maybe explicitly call out this assumption of purely IP addresses sans scheme and ports in the comment above the `disco.Addrs` call, for future reference ? Long as we have that documented, and an upgrade in go-discover version breaks that assumption, I think we should be fine.